### PR TITLE
Handle 401 with WWW-Authenticate. Moved wrong 401 into 400.

### PR DESCRIPTION
**Pull request by JonathanHuot**
OAuth2 Provider errors `access_denied`/`unauthorized_client`/`consent_required`/`login_required` MUST be 400, and not 401.

Also, the remaining 401 (`invalid_client` and `invalid_token`) MUST have WWW-Authenticate when set. It could have an impact of processing those in webframeworks.

Require an understanding of web frameworks (django, flask, ...) but also how themselves are implemented. Any feedback on this is more than welcome.

### CODE REVIEW

File: oauthlib/oauth2/rfc6749/endpoints/introspect.py [hunk 1]
```diff
@@ -63,6 +63,8 @@ def create_introspect_response(self, uri, http_method='POST', body=None,
             log.debug('Token introspect valid for %r.', request)
         except OAuth2Error as e:
             log.debug('Client error during validation of %r. %r.', request, e)
+            if e.status_code == 401:
+                return {"WWW-Authenticate": "Basic"}, e.json, e.status_code
```
[the remark this answers is not in the document]
Comment (JonathanHuot) [reply oauthlib/oauth2/rfc6749/endpoints/introspect.py]: Looks like Django-Oauth-Toolkit used the same `error=xx` solution, and merge it recently. See https://github.com/jazzband/django-oauth-toolkit/pull/584

[the remark this answers is not in the document]
Comment (skion) [reply oauthlib/oauth2/rfc6749/endpoints/introspect.py]: Yes, in favour of the alternative. BTW your current implementation is using `Basic` whereas the spec says:

> All challenges defined by this specification MUST use the auth-scheme
   value "Bearer"

File: oauthlib/oauth2/rfc6749/endpoints/introspect.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def create_introspect_response(self, uri, http_method='POST', body=None,
             log.debug('Token introspect valid for %r.', request)
         except OAuth2Error as e:
             log.debug('Client error during validation of %r. %r.', request, e)
-            return {}, e.json, e.status_code
+            return e.headers, e.json, e.status_code
```
[the remark this answers is not in the document]
Comment (JonathanHuot) [reply oauthlib/oauth2/rfc6749/endpoints/introspect.py]: Yeah, actually we could take this PR opportinity to add the Content-Type to the headers. It's also applicable for revocation endpoint. It was missing for both.

[the remark this answers is not in the document]
Comment (JonathanHuot) [reply oauthlib/oauth2/rfc6749/endpoints/introspect.py]: For `error` in both `WWW-Authenticate` and HTTP body, I have not seen any hard requirements, but the RFC specifies that 401 can be used, and `error` must be in the response's body. It doesn't said explicitly that both can be used in same time, though..

Introspect RFC:
```
   If the protected resource uses OAuth 2.0 client credentials to
   authenticate to the introspection endpoint and its credentials are
   invalid, the authorization server responds with an HTTP 401
   (Unauthorized) as described in Section 5.2 of OAuth 2.0 [RFC6749].
```

OAuth2 5.2 section:
```
   The authorization server responds with an HTTP 400 (Bad Request)
   status code (unless specified otherwise) and includes the following
   parameters with the response:
       error
         invalid_client
               (..).  The authorization server MAY
               return an HTTP 401 (Unauthorized) status code to indicate
               which HTTP authentication schemes are supported.  If the
               client attempted to authenticate via the "Authorization"
               request header field, the authorization server MUST
               respond with an HTTP 401 (Unauthorized) status code and
               include the "WWW-Authenticate" response header field
               matching the authentication scheme used by the client.

   (..)
   The parameters are included in the entity-body of the HTTP response
   using the "application/json" media type as defined by [RFC4627].  (..)
```

And then Bearer RFC6750:
```
3.  The WWW-Authenticate Response Header Field
   (..)
   All challenges defined by this specification MUST use the auth-scheme
   value "Bearer".  This scheme MUST be followed by one or more
   auth-param values.
   (..)
   If the protected resource request included an access token and failed
   authentication, the resource server SHOULD include the "error"
   attribute to provide the client with the reason why the access
   request was declined.
```

I think this 401/Authenticate topic deserves changes more in-depth, and this PR could be a short term solution to progress forward.
In the future, we should handle all areas of the (at least) three RFCs, mostly because we don't support everything (bearer authentication vs client authentication; sending `the authentication scheme used by the client`; `scope`/`realm`; and so on and so forth)


To summarize, examples in Bearer's RFC shows no HTTP body, the OAuth2's RFC
 require the error in the HTTP body. OAuthlib-2 send errors in the HTTP body, what should do OAuthlib-3 ? :-)
